packages/astro/src/core/build/common.ts

```diff
@@ -36,8 +37,15 @@ export function getOutFolder(
 					return new URL('.' + appendForwardSlash(pathname), outRoot);
 				}
 				case 'file': {
-					const d = pathname === '' ? pathname : npath.dirname(pathname);
-					return new URL('.' + appendForwardSlash(d), outRoot);
+					let dir;
+					// If the pathname is '' then this is the root index.html
+					// If this is an index route, the folder should be the pathname, not the parent
+					if(pathname === '' || routeData.index) {
```

Comment:
Nit: Maybe we could omit the check for an empty pathname since, in this scenario, it's guaranteed to represent an index route, isn't it?